useful_resources/sample_code/opt_outs_everyaction.py

```diff
@@ -0,0 +1,163 @@
+# This table's columns will be: vanid, phone, committeeid, and applied_at.
+SUCCESS_TABLE = os.environ['SUCCESS_TABLE']
+
+# The ERROR_TABLE is a schema.table where errors will be logged.
+# This table's columns will be : vanid, phone, committeeid, errored_at, and error.
+ERROR_TABLE = os.environ['ERROR_TABLE']
+
+# To use the Redshift connector, set the following environmental variables:
+# REDSHIFT_USERNAME
+# REDSHIFT_PASSWORD
+# REDSHIFT_HOST
+# REDSHIFT_DB
+# REDSHIFT_PORT
+
+rs = Redshift()
+
+
+def attempt_optout(ea, row, applied_at, committeeid, success_log, error_log, attempts_left=3):
+
+    vanid = row['vanid']
+    phone = row['phone']
+
+    # Documentation on this json construction is here
+    # https://docs.ngpvan.com/reference/common-models
+    json = {
+        "phones": [
+            {"phoneNumber": phone,
+             "phoneOptInStatus": "O"}
+        ]
+    }
+
+    try:
+        r = ea.update_person_json(id=vanid, match_json=json)
+
+        if isinstance(r, dict):
+            # Response is only a dict upon success
+            success_log.append({
+                "vanid": r.get('vanId'),
+                "phone": phone,
+                "committeeid": committeeid,
+                "applied_at": applied_at
+            })
+
+    # If we get an HTTP Error we log it
+    # Usually these errors are due to a vanid being deleted or merged
+    except requests.exceptions.HTTPError as e:
+        r = e
+        error_log.append({
+            "vanid": vanid,
+            "phone": phone,
+            "committeeid": committeeid,
+            "errored_at": applied_at,
+            "error": r
+        })
+
+    # If we get a connection error we wait a bit and try again.
+    except requests.exceptions.ConnectionError as c:
+        logger.info("Got disconnected, waiting and trying again")
+
+        while attempts_left > 0:
+            attempts_left -= 1
+
+            # Wait 10 seconds, then try again
+            time.sleep(10)
+            attempt_optout(ea, row, attempts_left)
+
+        else:
+            r = str(c)[:999]
+            return r
+
+            error_log.append({
+                "vanid": vanid,
+                "phone": phone,
+                "committeeid": committeeid,
+                "errored_at": applied_at,
+                "error": r
+            })
+            error_tbl = Table(error_log)
+            if error_tbl.num_rows() > 0:
+                error_tbl.to_redshift(
+                    ERROR_TABLE,
+                    if_exists='append',
+                    alter_table=True
+                )
+            raise Exception(f"Connection Error {r}")
+
+    return r
+
+
+def main():
+    # Creating empty lists where we'll log successes and errors
+    success_log = []
+    error_log = []
+
+    # Get the opt out data
+    s = rs.query(f"select * from {OPT_OUT_TABLE}")
+
+    # Turn the API keys into a list
+    API_KEYS = list(ast.literal_eval(API_KEYS_STR))
```

Comment:
I'll do a full review on Monday, but your question about the literal eval got me too curious to wait. 😂 What format is this string in? Does [json.loads()](https://www.geeksforgeeks.org/json-loads-in-python/#:~:text=loads()%20method%20can%20be,JSON%20data%20into%20Python%20Dictionary.) not work? If the issue is that it's not correct JSON because the two dicts aren't contained in anything, you could do something like:

`API_KEYS = json.loads("[" + API_KEYS_STR + "]")`

Still pretty hacky, but at least it avoids the literal_eval.
`

Reply:
Resolution: switch to `json.loads()` version, with a comment to the reader explaining what `json.loads()` is:

`API_KEYS = json.loads(API_KEYS_STR)`